events/api/src/main/java/org/projectnessie/events/api/CustomContent.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.api;
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@JsonTypeName("CUSTOM")
+@JsonSerialize
+@JsonDeserialize
+public interface CustomContent extends Content {
```

Comment:
Is this type supposed to report add actual content data as `Content.getAttributes()`? If so, could you add javadoc to clarify?

Reply:
Yes that's the intent. I will add some javadocs.

Reply:
I wonder whether we could have `Content.getAttributes()` simply duplicate all regular attributes _and_ allow it to include extra attributes... I guess that should make the handling `CustomContent` more straight-forward (i.e. avoid the `attributes` map in JSON). WDYT?

Reply:
> I wonder whether we could have Content.getAttributes() simply duplicate all regular attributes and allow it to include extra attributes...

I am not sure that this would make things easier tbh.

I have a code and tests in a separate branch that will handle this. Basically the mapping goes as follows:

```java
public static Content map(org.projectnessie.model.Content content) {
    org.projectnessie.model.Content.Type type = content.getType();
    if (type == org.projectnessie.model.Content.Type.NAMESPACE) {
      ...
    } else if (type == org.projectnessie.model.Content.Type.ICEBERG_TABLE) {
      ...
    } else if (type == org.projectnessie.model.Content.Type.ICEBERG_VIEW) {
      ...
    } else if (type == org.projectnessie.model.Content.Type.DELTA_LAKE_TABLE) {
      ...
    } else if (content instanceof GenericContent) {
      GenericContent genericContent = (GenericContent) content;
      ImmutableCustomContent.Builder builder =
          ImmutableCustomContent.builder()
              .id(UUID.fromString(Objects.requireNonNull(content.getId())))
              .customType(genericContent.getType().name());
      if (genericContent.getAttributes() != null) {
        builder.putAllAttributes(genericContent.getAttributes());
      }
      return builder.build();
    } else {
      Map<String, ?> map = MAPPER.convertValue(content, MAP_TYPE);
      return ImmutableCustomContent.builder()
          .id(UUID.fromString((String) map.remove("id")))
          .customType((String) map.remove("type"))
          .putAllAttributes(map)
          .build();
    }
  }
```

With that in mind, and assuming that `CustomAbcTable` is a non-built-in content type, e.g.:

```java
class CustomAbcTable extends org.projectnessie.model.Content {
      static final Type TYPE = new Type() {
            public String name() { return "CUSTOM_ABC_TABLE"; }
            public Class<? extends org.projectnessie.model.Content> type() { return CustomAbcTable.class; }
      };
      public Type getType() { return TYPE; }
      public String getId() { return UUID.randomUUID().toString();}
      public String getText() { return "foo"; }
      public int getNumber() { return 123; }
      public boolean isBoolean() { return true; }
}
```

Then the last `else` branch above will be used and will indeed produce a `CustomContent` that will be serialized as follows:

```json
{
  "type": "CUSTOM",
  "customType": "CUSTOM_ABC_TABLE",
  "id": "3171c949-b6f6-488a-9484-bd247c393223",
  "attributes": {
    "number": 123,
    "text": "foo",
    "boolean": true
  }
}
```

That is for non-built-in types. For the built-in ones, their serialization is straightforward and the `getAttributes()` map will be empty for now. Do you think we should duplicate their attributes there?

My thinking about placing the `getAttributes()` method in the parent `Content` interface was simply that it could be useful in the future if e.g. `IcebergTable` exposed a `getExtendedMetadata()` map or stg in the style.

Reply:
Update: the most recent changes modify slightly how a `CustomContent` is serialized; here is the updated example:

```json
{
  "type": "CUSTOM_ABC_TABLE",
  "id": "3171c949-b6f6-488a-9484-bd247c393223",
  "number": 123,
  "text": "foo",
  "boolean": true
}
```